communication/mavlink_waypoint_handler.cpp

```diff
@@ -517,1514 +385,285 @@ void Mavlink_waypoint_handler::clear_waypoint_list(Mavlink_waypoint_handler* way
+void Mavlink_waypoint_handler::advance_to_next_waypoint()
 {
-    mav_result_t result;
-
-    if ((waypoint_handler->navigation_.internal_state_ == Navigation::NAV_NAVIGATING) || (waypoint_handler->navigation_.internal_state_ == Navigation::NAV_HOLD_POSITION)
-        || (waypoint_handler->navigation_.internal_state_ == Navigation::NAV_STOP_ON_POSITION) || (waypoint_handler->navigation_.internal_state_ == Navigation::NAV_STOP_THERE))
+    // If the current waypoint index is the last waypoint, go to first waypoint
+    if (current_waypoint_index_ == (waypoint_count_-1))
     {
-        result = MAV_RESULT_ACCEPTED;
-
-        waypoint_handler->navigation_.auto_landing_behavior = Navigation::DESCENT_TO_SMALL_ALTITUDE;
-        waypoint_handler->auto_landing_next_state_ = false;
-
-        waypoint_handler->navigation_.internal_state_ = Navigation::NAV_LANDING;
-
-        //waypoint_handler->navigation_.dubin_state = DUBIN_INIT;
-
-        local_position_t landing_position = waypoint_handler->ins_.position_lf();
-        landing_position[Z] = -5.0f;
-        if (packet->param1 == 1)
-        {
-            print_util_dbg_print("Landing at a given location\r\n");
-
-            landing_position[X] = packet->param5;
-            landing_position[Y] = packet->param6;
-        }
-        else
-        {
-            print_util_dbg_print("Landing on the spot\r\n");
-        }
-
-        if (waypoint_handler->navigation_.navigation_strategy == Navigation::strategy_t::DUBIN)
-        {
-            waypoint_handler->dubin_hold_init(landing_position);
-        }
-        else
-        {
-            waypoint_handler->waypoint_hold_coordinates.position = landing_position;
-            waypoint_handler->waypoint_hold_coordinates.heading  = coord_conventions_get_yaw(waypoint_handler->ahrs_.qe);
-        }
-
-        print_util_dbg_print("Auto-landing procedure initialised.\r\n");
-
-        print_util_dbg_print("Landing at: (");
-        print_util_dbg_print_num(waypoint_handler->waypoint_hold_coordinates.position[X], 10);
-        print_util_dbg_print(", ");
-        print_util_dbg_print_num(waypoint_handler->waypoint_hold_coordinates.position[Y], 10);
-        print_util_dbg_print(", ");
-        print_util_dbg_print_num(waypoint_handler->waypoint_hold_coordinates.position[Z], 10);
-        print_util_dbg_print(", ");
-        print_util_dbg_print_num((int32_t)(waypoint_handler->waypoint_hold_coordinates.heading * 180.0f / 3.14f), 10);
-        print_util_dbg_print(")\r\n");
+        set_current_waypoint_index(0);
     }
-    else
+    else // Update current in both waypoints
     {
-        result = MAV_RESULT_DENIED;
+        set_current_waypoint_index(current_waypoint_index_+1);
     }
-
-    return result;
 }
 
-void Mavlink_waypoint_handler::auto_landing_handler()
+void Mavlink_waypoint_handler::control_time_out_waypoint_msg()
```

Comment:
Ok for all points except the last.
It may block if the ground station first asks the list of waypoint, then stops after a few waypoints, then decide to send a new waypoint list. That would be stupid but it could happen if the user clicks fast. I prefer to have the drone blindly follow the orders vs do additional checks for QGC.